File: c++/138. Copy List with Random Pointer.cpp
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
       // write your code here
        RandomListNode *p = head, *pre = nullptr;
        RandomListNode *new_head = nullptr;
        unordered_map<RandomListNode*, RandomListNode*> mp;
        while(p) {
            RandomListNode *node = new RandomListNode(p->label);
            if (pre) {
                pre->next = node;
            } else
                new_head = node;    
            mp[p] = node;
            pre = node;
            p = p->next;
        }
        
        p = head;
        RandomListNode *q = new_head;
        while(p && q) {
            q->random = mp[p->random];
            p = p->next;
            q = q->next;
        }
        
        return new_head;
    }
};

File: go/src/leetcode/138. Copy List with Random Pointer/138. Copy List with Random Pointer.go
package main

type Node struct {
	Val int
	Next *Node
	Random *Node
}

func copyRandomList(head *Node) *Node {
	var mapper = make(map[*Node]*Node)
	var copy func(node *Node) *Node
	copy = func(node *Node) *Node {
		if node == nil {
			return nil
		}
		if _, ok := mapper[node]; ok {
			return mapper[node]

		}
		mapper[node] = &Node{
			Val: node.Val,
		}
		c := mapper[node]
		c.Random = copy(node.Random)
		return c
	}

	var dummy = &Node{}
	var c = dummy
	for head != nil {
		c.Next = copy(head)
		head = head.Next
		c = c.Next
	}
	return dummy.Next
}
